beacon_node/eth1_chain/src/web3_fetcher.rs

```diff
@@ -0,0 +1,245 @@
+    fn get_deposit_logs_filter(&self) -> Filter {
+        /// Keccak256 hash of "DepositEvent" in bytes for passing to log filter.
+        const DEPOSIT_CONTRACT_HASH: &str =
+            "649bbc62d0e31342afea4e5cd82d4049e7e1ee912fc0889aa790803be39038c5";
+        let filter = FilterBuilder::default()
+            .address(vec![self.contract.address])
+            .topics(
+                Some(vec![DEPOSIT_CONTRACT_HASH.parse().unwrap()]),
+                None,
+                None,
+                None,
+            )
+            .build();
+        filter
+    }
+}
+
+impl Eth1DataFetcher for Web3DataFetcher {
+    /// Get block_number of current block.
+    fn get_current_block_number(&self) -> Option<U256> {
+        let block_future = self.web3.eth().block_number();
+        let block = block_future.wait().ok()?;
+        Some(block)
+    }
+
+    /// Get block hash at given height.
+    fn get_block_hash_by_height(&self, height: u64) -> Option<H256> {
+        let block_future = self
+            .web3
+            .eth()
+            .block(BlockId::Number(BlockNumber::Number(height)));
+        let block = block_future.wait();
+        block.ok().and_then(|x| x).and_then(|b| b.hash)
+    }
+
+    /// Get `deposit_count` from deposit contract at given eth1 block number.
+    fn get_deposit_count(&self, block_number: Option<BlockNumber>) -> Option<u64> {
+        let contract =
+            Contract::from_json(self.web3.eth(), self.contract.address, &self.contract.abi).ok()?;
+        let data = contract.query(
+            "get_deposit_count",
+            (),
+            None,
+            Options::default(),
+            block_number,
+        );
+        let deposit_count: Vec<u8> = data.wait().ok()?;
+        vec_to_u64_le(&deposit_count)
+    }
+
+    /// Get `deposit_root` from deposit contract at given eth1 block number.
+    fn get_deposit_root(&self, block_number: Option<BlockNumber>) -> Option<H256> {
+        let contract =
+            Contract::from_json(self.web3.eth(), self.contract.address, &self.contract.abi).ok()?;
+        let data = contract.query(
+            "get_hash_tree_root",
+            (),
+            None,
+            Options::default(),
+            block_number,
+        );
+        let deposit_root: Vec<u8> = data.wait().ok()?;
+        Some(H256::from_slice(&deposit_root))
+    }
+
+    /// Returns a future which subscribes to `DepositEvent` events and inserts the
+    /// parsed deposit into the passed cache structure everytime an event is emitted.
+    fn get_deposit_logs_subscription(
+        &self,
+        cache: Arc<RwLock<BTreeMap<u64, DepositData>>>,
+    ) -> Box<Future<Item = (), Error = ()> + Send> {
+        let filter: Filter = self.get_deposit_logs_filter();
+        let event_future = self
+            .web3
+            .eth_subscribe()
+            .subscribe_logs(filter)
+            .then(move |sub| {
+                sub.unwrap().for_each(move |log| {
```

Comment:
Is this `unwrap` safe?

If so, consider using an `expect()` with a description of why it's safe.

If not, handle gracefully and try log an error.